examples/hosts/app-integration/external-controller/src/app.ts

```diff
@@ -18,27 +21,24 @@ if (location.hash.length === 0) {
 const documentId = location.hash.substring(1);
 document.title = documentId;
 
+const dataObjectId = "dice";
+
 async function start(): Promise<void> {
-    // Get Fluid Container (creates if new url)
-    const container = await getTinyliciousContainer(documentId, KeyValueContainerRuntimeFactory, createNew);
-
-    // Since we're using a ContainerRuntimeFactoryWithDefaultDataStore, our dice roller is available at the URL "/".
-    const url = "/";
-    const response = await container.request({ url });
-
-    // Verify the response to make sure we got what we expected.
-    if (response.status !== 200 || response.mimeType !== "fluid/object") {
-        throw new Error(`Unable to retrieve data object at URL: "${url}"`);
-    } else if (response.value === undefined) {
-        throw new Error(`Empty response from URL: "${url}"`);
-    }
-
-    // In this app, we know our container code provides a default data object that is an IDiceRoller.
-    const keyValueDataObject: IKeyValueDataObject = response.value;
+    // Get or create the document
+    const fluidDocument = createNew
+        ? await Fluid.createDocument(documentId)
+        : await Fluid.getDocument(documentId);
+
+    // We'll create the data object when we create the new document.
+    const keyValueDataObject: IKeyValueDataObject = createNew
+        ? await fluidDocument.createDataObject<KeyValueDataObject>(KeyValueInstantiationFactory.type, dataObjectId)
+        : await fluidDocument.getDataObject<KeyValueDataObject>(dataObjectId);
```

Comment:
```suggestion
    let fluidDocument;
    if (createNew) {
        fluidDocument = await Fluid.createDocument(documentId);
        await fluidDocument.createDataObject<KeyValueDataObject>(KeyValueInstantiationFactory.type, dataObjectId);
    } else {
        fluidDocument = await Fluid.getDocument(documentId);
    }
    
    const keyValueDataObject = await fluidDocument.getDataObject<KeyValueDataObject>(dataObjectId);
```

Reply:
Even though I made this suggestion, I can't tell if having the createDataObject/getDataObject method is better than the request. I **do** like having the creation flow available through the request.

Reply:
I prefer hiding the request, but it will be good to get feedback on this.

Reply:
request is a two sided paradigm. Both the caller and the listener need to agree on the syntax.

It's been historically a very confusing section of our code because it invites complicated questions. Why am I requesting? What am I requesting? Why does the request go to the Container? What if I don't get the response I'm expecting? Why am I requesting "/"?

In non-dynamic scenarios we have the ability to encapsulate both sides of the request in a coherent way. In full stack scenarios I'm a huge proponent for the owner of the Container exposing functionality while using the request paradigm as an implementation detail of that functionality. IE. the developer does not want to request a "/" they want to get a DataObject.